color/clr
# Switch clr to argparse and inspect.Signature. Make bash completions work.

To test:
`pip install pip install git+https://github.com/color/clr.git@complete2  && source local/virtualenv3/lib/python3.6/site-packages/clr/completion.bash `

- optparse (deprecated) -> argparse
- inspect.getargspec (deprecated) -> inspect.Signature
- Show module doc strings in help
- Bash completions work for arguments.

Next (last?) up: zsh completions and integrating into color build env.

Used the opportunity to remove a bunch of code and improve help strings. 

Effort was made to keep arg parsing 100% compatible with previous version. Parsing is now more flexible, but all previous incantations should still work.


```
def cmd_argtest(self, a, b, c=4, d=None, e=False):
        """For testing arg parsing."""
        print(a, b, c, d, e)
```
Before:
```
$ clr help argtest
argtest <a> <b> [--c=4 --d=None --e]
  For testing arg parsing.
$ clr argtest 1 2 
1 2 4 None False
$ clr argtest 1 2 3
1 2 3 None False
$ clr argtest 1 2 3 4
1 2 3 4 False
$ clr argtest 1 2 3 4 --e
1 2 3 4 True
$ clr argtest 1 2 --c=3
1 2 3 None False
$ clr argtest 1 2 --c=3 --d=4
1 2 3 4 False
$ clr argtest 1 2 --c=3 --d 4
1 2 3 4 False
$ clr argtest 1 2 --c 3 
1 2 3 None False
$ clr argtest 1
Not all non-default arguments were specified!
argtest <a> <b> [--c=4 --d=None --e]
  For testing arg parsing.
$ clr argtest 1 --b=2
Not all non-default arguments were specified!
argtest <a> <b> [--c=4 --d=None --e]
  For testing arg parsing.
$ clr argtest --a=1 --b=2
Not all non-default arguments were specified!
argtest <a> <b> [--c=4 --d=None --e]
  For testing arg parsing.
$ clr argtest 1 2 --a 3 
1 2 4 None False
```

After:
```
$ clr help argtest
usage: clr system:argtest [--a A] [--b B] [--c C] [--d D] [--e | --noe]
                          [a] [b] [c] [d]

For testing arg parsing.

positional arguments:
  a      Required. Can also be specified with --a.
  b      Required. Can also be specified with --b.
  c      Optional. Defaults to c='4'. Can also be specified with --c.
  d      Optional. Defaults to d='None'. Can also be specified with --d.

optional arguments:
  --a A  Required. Can also be specified with positional arg a.
  --b B  Required. Can also be specified with positional arg b.
  --c C  Optional. Defaults to c='4'. Can also be specified with positional
         arg c.
  --d D  Optional. Defaults to d='None'. Can also be specified with positional
         arg d.
  --e    Optional. Defaults to e='False'.
  --noe  Optional. Defaults to e='False'.
$ clr argtest 1 2
1 2 4 None False
$ clr argtest 1 2 3
1 2 3 None False
$ clr argtest 1 2 3 4
1 2 3 4 False
$ clr argtest 1 2 3 4 --e
1 2 3 4 True
$ clr argtest 1 2 3 4 --noe
1 2 3 4 False
$ clr argtest 1 2 3 4 --noe --e
usage: clr system:argtest [--a A] [--b B] [--c C] [--d D] [--e | --noe]
                          [a] [b] [c] [d]
clr system:argtest: error: argument --e: not allowed with argument --noe
$ clr argtest 1 2 --c=3
1 2 3 None False
$ clr argtest 1 2 --c 3
1 2 3 None False
$ clr argtest 1 2 --c=3 --d=4
1 2 3 4 False
$ clr argtest 1 2 --c=3 --d=4 --e
1 2 3 4 True
$ clr argtest 1 2 --c=3 --d=4 --noe
1 2 3 4 False
$ clr argtest 1
usage: clr system:argtest [--a A] [--b B] [--c C] [--d D] [--e | --noe]
                          [a] [b] [c] [d]
clr system:argtest: error: one of the arguments --b b is required
$ clr argtest 1 --b=2
1 2 4 None False
$ clr argtest 1 --a=2
usage: clr system:argtest [--a A] [--b B] [--c C] [--d D] [--e | --noe]
                          [a] [b] [c] [d]
clr system:argtest: error: argument --a: not allowed with argument a
$ clr argtest --a=1 --b=2
1 2 4 None False
$ clr argtest --a=1 --b=2 4
usage: clr system:argtest [--a A] [--b B] [--c C] [--d D] [--e | --noe]
                          [a] [b] [c] [d]
clr system:argtest: error: argument a: not allowed with argument --a
$ clr argtest --a=1 --b=2 --c=3 --d=4
1 2 3 4 False
$ clr <TAB><TAB>
ancestry:          clinical_support:  content:           data_return:       elasticsearch:     lambda:            pipeline_ops:      src:               usage              web:
argtest            clinvar:           counseling:        datasets:          etl:               load_test:         product:           support:           usd:               wisdom:
audit:             completion1        covid:             deploy:            help               perf:              profile_imports    swagger:           validation:        
census:            completion2        cypress:           django:            js:                pg:                reports:           system:            variants:          
clear_cache        confirmation:      data:              ec2:               keyfile:           pipeline:          s3:                test_fixtures:     vcf:               
$ clr ar<TAB>gtest <TAB>--<TAB><TAB>
--a     --b     --c     --d     --e     --noe   
```
